staging/src/k8s.io/client-go/tools/watch/retrywatcher.go

```diff
@@ -268,7 +268,13 @@ func (rw *RetryWatcher) receive() {
 			return
 		}
 
-		time.Sleep(retryAfter)
+		timer := time.NewTimer(retryAfter)
+		defer timer.Stop()
+		select {
+		case <-ctx.Done():
+			return
+		case <-timer.C:
+		}
```

Comment:
select {
case <- time.After(retryAdfter):
case <-ctx.Done():
  return
}

Reply:
Yes. But the code in the PR is written like that on purpose because `time.After` creates a timer that is not freed until it expires, even if it's not needed it's not garbage collected. If we create the timer ourself, we can call `Stop` on it to free up the resources.

https://www.arangodb.com/2020/09/a-story-of-a-memory-leak-in-go-how-to-properly-use-time-after/

Hm, and this article suggests a different pattern. I'm going to look into it now...

Reply:
It's not clear whether it's necessary to drain the channel to avoid the memory leak or it's just needed to prevent receives from it after stop has been called https://github.com/golang/go/blob/master/src/time/sleep.go#L59-L65

Reply:
Ok, looks like what I'm suggesting in this PR is the right way to do it https://github.com/golang/go/issues/27169

Reply:
See also https://github.com/golang/go/issues/37196